README.md
# cloudify-netconf-plugin
Cloudify plugin for serializing TOSCA node templates to netconf configuration.

Support:
* both version of netconf protocol from rfc6242.
* ssh based communication

Code for now does not support:
* conditional statements with dependencies capabilities available on hardware level,
* can be described only input structures for netconf, output(received structures from netconf) does not validated and not described in yaml blueprints.
* complicated structures with list of types can be described only as static(not dynamic) in sophisticated cases
* case when we have to different structures by some condition, like we send configuration that in case dhcp - contain only flag that we have dhcp configuration,
otherwise we have full structures that have described static connection. So if in description we have both fields, in first case we send empty values, so we need to have
2 node types/configurations for dhcp and for static

For check generation:
* install tools/setup.py and yttc from https://github.com/cloudify-cosmo/yttc
* xml to yaml: netconfxml2yaml.py cloudify-netconf-plugin/tools/examples/rpc.xml
* yaml to xml: yaml2netconfxml.py cloudify-netconf-plugin/tools/examples/rpc.yaml
* generate parent yaml blueprint for include:
  cd tools/examples
  yttc turing-machine.yang

Vyatta example is valid only for Brocade Vyatta Network OS 4.1 R2 and before run vyatta blueprint run as root on router:
* cd /usr/share/doc/openvpn/examples/sample-keys/
* bash gen-sample-keys.sh

Script name can be different and related to Brocade vRouter version.

## tags name conversions logic:
* ```a``` -> tag with name "a" and namespaces will be same as parent
* ```a@b``` -> tag with name "b" and namespace a
* ```_@@a``` -> attribute with name a and namespace will be same as parent
* ```_@a@b``` -> attribute with name b and namespace will be a
* ```_@@``` -> text content for tag
* ```_!_``` -> text will be inserted as xml to parent node

## examples of conversion

### list
from:
```
{
    "b": {
        "a": [1, 2, 3],
        "c": 4
    }
}
```
to:
```
<b>
    <a>1</a>
    <a>2</a>
    <a>3</a>
    <c>4</c>
</b>
```
### dict
from:
```
{
    "b": {
        "a": 1,
        "c": 2
    }
}
```
to:
```
<b>
    <a>1</a>
    <c>2</c>
</b>
```
### attributes
from:
```
{
    "b": {
        "_@@a": 1,
        "_@@": 2
    }
}
```
to:
```
<b a=1>
    2
</b>
```
### text value for tag with attibutes
from:
```
{
    "b@a": {
        "_@c@a": 1,
        "_@@": 2,
        "_@@g": 3
    }
}
```
to:
```
<b:a c:a=1 b:g=3>
    2
</b:a>
```

### text value for raw insert
from:
```
{
    "a": {
        "_!_": "<g><a></a><c></c><d></d></g>"
    }
}
```
to:
```
<a>
    <g>
        <a/>
        <c/>
        <d/>
    </g>
</a>
```

## Node description example
```

node_templates:
  some-implimentation:
    type: cloudify.netconf.nodes.xml_rpc
    properties:
      netconf_auth:
        user: <user on device>
        password: <password on device, optional>
        ip: <ip for device or list of ip's if have failback ip's>
        key_content: <private key content>
        port: <port for device, optional by default 830>
        store_logs: <store all communication logs, optional by default false>
      metadata:
        xmlns:
          _: <default namespace>
        capabilities:
          <list of capabilities, optional>
    interfaces:
      cloudify.interfaces.lifecycle:
        create: // can be create / configure / start / stop / delete
          inputs:
            lock:
              <list databased for lock, optional>
            back_database: <datebase to copy, optional>
            front_database: <database for copy, optional>
            strict_check: <ignore not critical errors in xml, optional>
            calls:
              - action: <action name for netconf>
                payload:
                  <dict for put to payload>
                save_to: <field name for save to runtime properties, optional>
                deep_error_check: <look deeply for errors, optional by default false>
```

## Node description example with xml template

### xml command list file part (template file)
```
<?xml version='1.0' encoding='UTF-8'?>
<rfc6020:rpc xmlns:rfc6020="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:xnm="http://xml.juniper.net/xnm/1.1/xnm" rfc6020:message-id="1001">
  <rpc-command-one/>
</rfc6020:rpc>]]>]]>

<?xml version='1.0' encoding='UTF-8'?>
<rfc6020:rpc xmlns:rfc6020="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:xnm="http://xml.juniper.net/xnm/1.1/xnm" rfc6020:message-id="1002">
  <rpc-command-two>
    {{ some_param_name }}
  </rpc-command-two>
</rfc6020:rpc>]]>]]>
```
As separator between commands must be used ```]]>]]>```

### yaml example part
```

node_templates:
  some-implimentation:
    type: cloudify.netconf.nodes.xml_rpc
    properties:
      netconf_auth:
        user: <user on device>
        password: <password on device, optional>
        ip: <ip for device>
        key_content: <private key content>
        port: <port for device, optional by default 830>
      metadata:
        xmlns:
          _: <default namespace>
        capabilities:
          <list of capabilities, optional>
    interfaces:
      cloudify.interfaces.lifecycle:
        create: // can be create / configure / start / stop / delete
          inputs:
            strict_check: <ignore not critical errors in xml, optional>
            deep_error_check: <look deeply for errors, optional by default false>
            template: <template file name or URL>
            params: <dict params for template, optional>
              some_param_name: some param name
```

_template_ input suppoorts single file location. It can be:
* a path which is resolverd as blueprint resource
* a URL

```

node_templates:
  some-implimentation:
    type: cloudify.netconf.nodes.xml_rpc
    properties:
      netconf_auth:
        user: <user on device>
        password: <password on device, optional>
        ip: <ip for device>
        key_content: <private key content>
        port: <port for device, optional by default 830>
      metadata:
        xmlns:
          _: <default namespace>
        capabilities:
          <list of capabilities, optional>
    interfaces:
      cloudify.interfaces.lifecycle:
        create: // can be create / configure / start / stop / delete
          inputs:
            strict_check: <ignore not critical errors in xml, optional>
            deep_error_check: <look deeply for errors, optional by default false>
            templates: [<template file name or URL>]
            params: <dict params for template, optional>
              some_param_name: some param name
```

_templates_ input may also be used instead of a single _template_. _templates_ does
not support multiple calls in one file - each call sould be provided as a
separete call.
The same ways of specifying file location as for template are supported:
blueprint resource or URL.

Notes:

* All content in metadata in common usage generated by yttc command, and
you can use output of such tool for 'parent/template' blueprint that can
be base of yours blueprint. Its not strictly needed but very useful
because you can use results of such tools as base for validate your
blueprint.
* You have examples for all basic operation of netconf protocol in
properties of base class of node (described in plugin.yaml) so you can
use all of such operation by `payload: { get_property: [ your_node, rfc6020@unlock ] }`.
* If you feel self as confidential person - you can write any structure
to payload without any validation by in yang. Sometime it will
be very useful for crazy devices without correct yang schemas.
* Also we have magic help functionality like `lock`, `back_database`,
`front_database`. So one by one: `lock` - we will lock database in list
for you before any operation and unlock after operations, its very useful
if you have several consumers that want to change configuration and
doesn't like to sync before change. `back_database`/`front_database` -
we can copy configuration from front database to back database before run
operations and all operations we will do in this database(if you don't
set any in rfc6020@target) and after will move your changes to front database.
Such functionality is useful for case when you want to have consistent
settings with apply after all changes. And sometime device does not
support change configuration in place.
* `save_to` - response from server will be saved to runtime node properties
files with such name and namespaces for this result will be saved to field
with same name and suffix `_ns`.
* We have copy of xslt/relaxng files from `pyang` in `share-files` and have
fully same license as original files from pyang.
* you can skip set `ip` in `netconf_auth` in case if your node have
`cloudify.nodes.Compute` node in relationships with type
`cloudify.relationships.contained_in`. In such case as `ip` will be used
`ip` from compute node.
* as value `ip` in `netconf_auth` can be list of `ip's` if you have failback.

## Versions:

Look to [ChangeLog](CHANGELOG.txt).

## Examples
For official blueprint examples using this Cloudify plugin, please see [Cloudify Community Blueprints Examples](https://github.com/cloudify-community/blueprint-examples/).


CHANGELOG.txt
releases:
0.4.12: Release DSL 1_5 and manylinux wagon.
0.4.11: Release DSL 1_4 and redhat 8 wagon.
0.4.10: dummy test release.
0.4.9: dummy test release.
0.4.8: dummy test release.
0.4.7: dummy test release.
0.4.6: dummy test release.
0.4.5: dummy test release.
0.4.4:
- Add release management to Circleci and Update wagon builder to py2py3 wagon.

0.4.3:
- Port the plugin to Python 3

0.4.2:
- Extracted main entry method to allow binding with relationship operations
- Use cloudify-utilities-plugins-sdk==0.0.10 from PYPI

0.4.1:
- Reset candidate database on failure
- Close connection to service after failure

0.4.0:
- Bump dependencies.
- Support list of url templates.
- Deprecate dsdl validation.
- Deprecate xml2netconf node type, use templates instead.
- Support rerun operation on resume.

0.3.1:
- Use cloudify-utilities-plugins-sdk==0.0.2 for manage netconf low level connection.
- Support early close connection before we fully read message.

0.3.0:
- Use cloudify-utilities-plugins-sdk==0.0.1 for manage netconf low level connection.

0.2.1:
- Ignore comments in xml
- ignore namespaces in erorr processing

0.2.0:
- Juniper devices support
- Cisco devices support
- Provide ctx in template
- Add "raw insert" to xml
- Suport ip list failback

0.1.0: Initial release


cloudify_netconf/xml_rpc.py
# Copyright (c) 2015-2019 Cloudify Platform Ltd. All rights reserved
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from lxml import etree
import time
import requests

from cloudify_common_sdk import exceptions
import cloudify_terminal_sdk.netconf_connection as netconf_connection
from cloudify.constants import NODE_INSTANCE, RELATIONSHIP_INSTANCE
from cloudify.decorators import operation
from cloudify import exceptions as cfy_exc
from cloudify_common_sdk import filters
import cloudify_netconf.utils as utils

from cloudify_netconf._compat import (text_type, urlparse)


def _generate_hello(xmlns, netconf_namespace, capabilities):
    """generate initial hello message with capabilities"""
    if not capabilities:
        capabilities = []
    if netconf_connection.NETCONF_1_0_CAPABILITY not in capabilities:
        capabilities.append(netconf_connection.NETCONF_1_0_CAPABILITY)
    if netconf_connection.NETCONF_1_1_CAPABILITY not in capabilities:
        capabilities.append(netconf_connection.NETCONF_1_1_CAPABILITY)
    hello_dict = {
        netconf_namespace + '@capabilities': {
            netconf_namespace + '@capability': capabilities
        }
    }

    hello_xml = utils.generate_xml_node(
        hello_dict,
        xmlns,
        netconf_namespace + '@hello'
    )
    return etree.tostring(
        hello_xml, pretty_print=True, xml_declaration=True,
        encoding='UTF-8'
    )


def _generate_goodbye(xmlns, netconf_namespace, message_id):
    """general final goodbye message"""
    goodbye_dict = {
        netconf_namespace + '@close-session': None,
        '_@@message-id': message_id
    }
    goodbye_xml = utils.generate_xml_node(
        goodbye_dict,
        xmlns,
        netconf_namespace + '@rpc'
    )
    return etree.tostring(
        goodbye_xml, pretty_print=True, xml_declaration=True,
        encoding='UTF-8'
    )


def _server_support_1_1(xmlns, netconf_namespace, response):
    xml_node = etree.XML(response)
    xpath = (
        "/%s:hello/%s:capabilities/%s:capability" % (
            netconf_namespace, netconf_namespace, netconf_namespace
        )
    )
    capabilities = xml_node.xpath(xpath, namespaces=xmlns)
    for node in capabilities:
        xml_dict = utils.generate_dict_node(node, xmlns)
        value = xml_dict.get(netconf_namespace + '@capability')
        if value == netconf_connection.NETCONF_1_1_CAPABILITY:
            return True
    return False


def _have_error(reply):
    # error check
    # https://tools.ietf.org/html/rfc6241#section-4.3
    error = None

    errors = []
    for k in reply:
        # skip attributes
        if k[:2] == "_@":
            continue
        if 'rpc-error' == k or '@rpc-error' == k[-len('@rpc-error'):]:
            # we can have list of rpc-errors with different namespaces
            if not isinstance(reply[k], list):
                # only one error with such namespace
                errors.append(reply[k])
            else:
                # we can combine several errors lists
                errors += reply[k]

    if not errors:
        return

    for error in errors:
        if not error:
            # we have empty rpc-error?
            raise cfy_exc.RecoverableError(
                "Empty error struct:" + str(errors)
            )

        # server can send warning as error, lets check
        error_severities = []
        for k in error:
            # skip attributes
            if k[:2] == "_@":
                continue
            if (
                'error-severity' == k or
                '@error-severity' == k[-len('@error-severity'):]
            ):
                # and error severity that we found
                error_severities.append(error[k])
        for error_severity in error_severities:
            if error_severity != 'warning':
                raise cfy_exc.RecoverableError(
                    "We have error in reply: {}".format(repr(errors)))


def _search_error(reply, netconf_namespace):
    # recursive search for error tag, slow and dangerous
    if isinstance(reply, text_type):
        return
    elif isinstance(reply, list):
        for tag in reply:
            _search_error(tag, netconf_namespace)
    elif isinstance(reply, dict):
        _have_error(reply)
        for tag_name in reply:
            _search_error(reply[tag_name], netconf_namespace)
            if tag_name.find("@rpc-error") != -1 and tag_name[:2] != "_@":
                # repack to detect error with different namespace
                _have_error({'rpc-error': reply[tag_name]})


def _check_reply_for_errors(reply, netconf_namespace, deep_error_check=False):
    if deep_error_check:
        # only for case when we have errors in the middle of message
        # is not described in https://tools.ietf.org/html/rfc6241#section-4.3
        # but can be in wild life
        _search_error(reply, netconf_namespace)
    else:
        _have_error(reply)

    return reply


def _parse_response(ctx, xmlns, netconf_namespace, response,
                    strict_check=False, deep_error_check=False):
    """parse response from server with check to rpc-error"""
    if strict_check:
        try:
            xml_node = etree.XML(response)
        except etree.XMLSyntaxError as e:
            raise cfy_exc.NonRecoverableError(
                "Syntax error in xml %s" % str(e)
            )
    else:
        # for case when we recieved not fully correct xml
        parser = etree.XMLParser(recover=True)
        xml_node = etree.XML(response, parser)

    xml_dict = utils.generate_dict_node(xml_node, xmlns)

    try:
        if 'rpc-reply' not in xml_dict and \
                (netconf_namespace + '@rpc-reply') not in xml_dict:
            ctx.logger.error(
                'Unexpected key in response: {response}'.format(
                    response=filters.shorted_text(xml_dict)))
        reply = \
            [v for k, v in xml_dict.items()
             if 'rpc-reply' in k][0]
    except IndexError:
        raise cfy_exc.NonRecoverableError(
            'unexpected reply struct: {0}'.format(xml_dict)
        )

    _check_reply_for_errors(reply, netconf_namespace, deep_error_check)

    return reply


def _merge_ns(base, override):
    """we can have several namespaces in properties and in call"""
    new_ns = {}
    for ns in base:
        new_ns[ns] = base[ns]

    for ns in override:
        new_ns[ns] = override[ns]

    return new_ns


def _run_one_string(ctx, netconf, rpc_string, xmlns, netconf_namespace,
                    strict_check, deep_error_check):
    ctx.logger.info(
        "Checks: xml validation: {strict_check}, "
        "rpc_error deep check: {deep_error_check} "
        .format(strict_check=strict_check,
                deep_error_check=deep_error_check)
    )
    ctx.logger.debug("Sent: {message}"
                     .format(message=filters.shorted_text(rpc_string)))

    # cisco send new line before package, so need strip
    try:
        response = netconf.send(rpc_string).strip()
    except exceptions.NonRecoverableError as e:
        # use str instead, for fully hide traceback and orignal exception name
        raise cfy_exc.NonRecoverableError(str(e))

    ctx.logger.debug("Recieved: {response}"
                     .format(response=filters.shorted_text(response)))

    response_dict = _parse_response(
        ctx, xmlns, netconf_namespace, response, strict_check, deep_error_check
    )
    ctx.logger.debug("Package: {response}"
                     .format(response=filters.shorted_text(response_dict)))
    return response_dict


def _run_one(ctx, netconf, message_id, operation, netconf_namespace, data,
             xmlns, strict_check=False, deep_error_check=False):
    """run one call by netconf connection"""
    # rpc
    ctx.logger.info("call: {call}".format(call=operation))
    parent = utils.rpc_gen(
        message_id, operation, netconf_namespace, data, xmlns
    )

    # send rpc
    rpc_string = etree.tostring(
        parent, pretty_print=True, xml_declaration=True,
        encoding='UTF-8'
    )

    return _run_one_string(ctx, netconf, rpc_string, xmlns, netconf_namespace,
                           strict_check, deep_error_check)


def _lock(ctx, name, lock, netconf, message_id, netconf_namespace, xmlns,
          strict_check):
    """lock database by name"""
    operation = "@lock" if lock else "@unlock"
    data = {
        netconf_namespace + "@target": {
            name: {}
        }
    }
    _run_one(
        ctx, netconf, message_id, netconf_namespace + operation,
        netconf_namespace, data, xmlns, strict_check
    )


def _discard_changes(ctx, netconf, message_id, netconf_namespace, xmlns,
                     strict_check):
    """discard changes in candidate database"""
    operation = "@discard-changes"
    _run_one(
        ctx, netconf, message_id, netconf_namespace + operation,
        netconf_namespace, {}, xmlns, strict_check
    )


def _copy(ctx, front, back, netconf, message_id, netconf_namespace, xmlns,
          strict_check):
    """copy fron database values to back database"""
    data = {
        netconf_namespace + "@source": {
            front: {}
        },
        netconf_namespace + "@target": {
            back: {}
        }
    }
    _run_one(
        ctx, netconf, message_id, netconf_namespace + "@copy-config",
        netconf_namespace, data, xmlns, strict_check
    )


def _update_data(data, operation, netconf_namespace, back):
    """update operation with database name"""
    if operation != 'rfc6020@edit-config':
        return
    if not back:
        return
    if "target" in data:
        if not data["target"]:
            data["target"] = {back: None}
        return
    if netconf_namespace + "@target" in data:
        if data[netconf_namespace + "@target"]:
            return
    data[netconf_namespace + "@target"] = {back: None}


def _run_templates(ctx, netconf, templates, template_params, netconf_namespace,
                   xmlns, strict_check, deep_error_check):
    for template in templates:
        # initially empty
        if not template:
            continue

        template = template.strip()
        # empty after strip
        if not template:
            continue

        if not template_params:
            template_params = {}

        # supply ctx for template for reuse runtime params
        template_params['ctx'] = ctx
        rpc_string = filters.render_template(template, template_params)

        _run_one_string(ctx, netconf, rpc_string, xmlns, netconf_namespace,
                        strict_check, deep_error_check)


def _run_calls(ctx, netconf, message_id, netconf_namespace, xmlns, calls,
               back_database, strict_check):
    # we can have several calls in one session,
    # like lock, edit-config, unlock
    for call in calls:
        deep_error_check = call.get('deep_error_check')
        operation = call.get('action')
        if not operation:
            ctx.logger.info("No operations")
            continue
        data = call.get('payload', {})

        message_id += 1

        if "@" not in operation:
            operation = "_@" + operation

        _update_data(data, operation, netconf_namespace, back_database)

        response_dict = _run_one(
            ctx, netconf, message_id, operation, netconf_namespace, data,
            xmlns, strict_check, deep_error_check
        )

        # save results to runtime properties
        save_to = call.get('save_to')
        if save_to:
            instance = None
            if ctx.type == NODE_INSTANCE:
                instance = ctx.instance
            elif ctx.type == RELATIONSHIP_INSTANCE:
                instance = ctx.source.instance

            if instance is not None:
                instance.runtime_properties[save_to] = response_dict
                instance.runtime_properties[save_to + "_ns"] = xmlns
            else:
                ctx.logger.error('Cannot find instance to save response to')


def _get_template(ctx, template_location):
    parse_result = urlparse(template_location)
    if all([parse_result.scheme, parse_result.path]):
        if parse_result.scheme == 'file':
            with open(parse_result.path) as tmpl_f:
                return tmpl_f.read()
        else:
            return requests.get(template_location).content
    else:
        return ctx.get_resource(template_location)


def _run_in_database(ctx, netconf, message_id, netconf_namespace, xmlns, calls,
                     templates, kwargs, strict_check):
    """Change current database and run action"""
    if 'back_database' in kwargs and 'front_database' in kwargs:
        message_id += 1
        _copy(
            ctx, kwargs['front_database'], kwargs['back_database'],
            netconf, message_id, netconf_namespace, xmlns, strict_check
        )

    try:
        if calls:
            _run_calls(ctx, netconf, message_id, netconf_namespace, xmlns,
                       calls, kwargs.get('back_database'), strict_check)
        elif templates:
            template_params = kwargs.get('params')
            deep_error_check = kwargs.get('deep_error_check')
            ctx.logger.debug("Params for template {template_params}".format(
                template_params=filters.shorted_text(template_params)))
            _run_templates(ctx, netconf, templates, template_params,
                           netconf_namespace, xmlns, strict_check,
                           deep_error_check)

        if 'back_database' in kwargs and 'front_database' in kwargs:
            message_id += 1
            _copy(
                ctx, kwargs['back_database'], kwargs['front_database'],
                netconf, message_id, netconf_namespace, xmlns, strict_check
            )

    except (cfy_exc.NonRecoverableError, cfy_exc.RecoverableError) as e:
        # discard only if we know that used candidate database
        if 'back_database' in kwargs and 'front_database' in kwargs:
            ctx.logger.info("Discard changes")
            message_id += 1
            _discard_changes(ctx, netconf, message_id, netconf_namespace,
                             xmlns, strict_check)
        raise e
    return message_id


def _run_in_locked(ctx, netconf, message_id, netconf_namespace, xmlns, calls,
                   templates, kwargs, strict_check):
    """Run actions in locked state"""
    if 'lock' in kwargs:
        for name in kwargs['lock']:
            message_id += 1
            _lock(
                ctx, name, True, netconf, message_id, netconf_namespace,
                xmlns, strict_check
            )
    try:
        message_id = _run_in_database(
            ctx=ctx, netconf=netconf, message_id=message_id,
            netconf_namespace=netconf_namespace, xmlns=xmlns,
            calls=calls, templates=templates, kwargs=kwargs,
            strict_check=strict_check)
    finally:
        # unlock databases
        if 'lock' in kwargs:
            for name in kwargs['lock']:
                message_id += 1
                _lock(
                    ctx, name, False, netconf, message_id, netconf_namespace,
                    xmlns, strict_check
                )
    return message_id


@operation(resumable=True)
def run(ctx, **kwargs):
    """main entry point for instance lifecycle calls"""
    netconf_auth = ctx.node.properties.get('netconf_auth', {})

    # if node contained in some other node, try to overwrite ip
    if not netconf_auth.get('ip'):
        netconf_auth['ip'] = ctx.instance.host_ip
        ctx.logger.info("Used host from container: {ip_list}".format(
            ip_list=filters.shorted_text(netconf_auth['ip'])))

    metadata = ctx.node.properties.get('metadata', {})
    base_xmlns = ctx.node.properties.get('base_xmlns', {})
    run_with_properties(ctx,
                        netconf_auth,
                        metadata,
                        base_xmlns,
                        **kwargs)


@operation(resumable=True)
def run_with_properties(ctx,
                        netconf_auth,
                        metadata,
                        base_xmlns,
                        log_file_name=None,
                        **kwargs):
    """main entry point for calls with custom configuration"""

    calls = kwargs.get('calls', [])

    templates_locs = kwargs.get('templates', [])
    template = kwargs.get('template')

    templates = []
    for tmpl_loc in templates_locs:
        templates.append(_get_template(ctx, tmpl_loc))

    if template:
        templates.extend(_get_template(ctx, template).split("]]>]]>"))

    if not calls and not templates:
        ctx.logger.info("Please provide calls or template")
        return

    # credentials
    netconf_auth.update(kwargs.get('netconf_auth', {}))
    user = netconf_auth.get('user')
    password = netconf_auth.get('password')
    key_content = netconf_auth.get('key_content')
    port = int(netconf_auth.get('port', 830))
    ip_list = netconf_auth.get('ip')
    if isinstance(ip_list, text_type):
        ip_list = [ip_list]
    # save logs to debug file
    if netconf_auth.get('store_logs'):
        if log_file_name is None:
            if ctx.type == NODE_INSTANCE:
                log_file_name = (
                    "/tmp/netconf-"
                    "{execution_id}_{instance_id}_{workflow_id}.log"
                    .format(execution_id=str(ctx.execution_id),
                            instance_id=str(ctx.instance.id),
                            workflow_id=str(ctx.workflow_id))
                )
            elif ctx.type == RELATIONSHIP_INSTANCE:
                log_file_name = (
                    "/tmp/netconf-"
                    "{execution_id}_{src_instance_id}_"
                    "{trg_instance_id}_{workflow_id}.log"
                    .format(execution_id=str(ctx.execution_id),
                            src_instance_id=str(ctx.source.instance.id),
                            trg_instance_id=str(ctx.target.instance.id),
                            workflow_id=str(ctx.workflow_id))
                )
            log_file_name = (
                "/tmp/netconf-{execution_id}_{workflow_id}.log"
                .format(execution_id=str(ctx.execution_id),
                        workflow_id=str(ctx.workflow_id))
            )
        ctx.logger.info(
            "Communication logs will be saved to {log_file_name}".format(
                log_file_name=log_file_name)
        )

    strict_check = kwargs.get('strict_check', True)

    # check minimal amout of credentials
    if not port or not ip_list or not user or (
        not password and not key_content
    ):
        raise cfy_exc.NonRecoverableError(
            "please check your credentials"
        )

    # some random initial message id, for have different between calls
    message_id = int((time.time() * 100) % 100 * 1000)

    # xml namespaces and capabilities
    xmlns = metadata.get('xmlns', {})

    # override by system namespaces
    xmlns = _merge_ns(xmlns, base_xmlns)
    netconf_namespace, xmlns = utils.update_xmlns(
        xmlns
    )
    capabilities = metadata.get('capabilities')

    # connect
    ctx.logger.info("use {user}@{ip_list}:{port} for login".format(
        user=user, ip_list=ip_list, port=port))
    hello_string = _generate_hello(
        xmlns, netconf_namespace, capabilities
    ).decode('utf-8')
    ctx.logger.debug("Sent: {message}"
                     .format(message=filters.shorted_text(hello_string)))

    netconf = netconf_connection.NetConfConnection(logger=ctx.logger,
                                                   log_file_name=log_file_name)
    for ip in ip_list:
        try:
            capabilities = netconf.connect(
                ip, user, hello_string, password, key_content, port
            )
            ctx.logger.info("Will be used: {ip}".format(ip=ip))
            break
        except Exception as ex:
            ctx.logger.info("Can't connect to {ip} with {ex}".format(
                ip=repr(ip), ex=str(ex)
            ))
    else:
        raise cfy_exc.NonRecoverableError(
            "please check your ip list"
        )

    ctx.logger.debug("Recieved: {capabilities}"
                     .format(capabilities=filters.shorted_text(capabilities)))

    if _server_support_1_1(xmlns, netconf_namespace, capabilities):
        ctx.logger.info("use version 1.1 of netconf protocol")
        netconf.current_level = netconf_connection.NETCONF_1_1_CAPABILITY
    else:
        ctx.logger.info("use version 1.0 of netconf protocol")

    try:
        message_id = _run_in_locked(
            ctx=ctx, netconf=netconf, message_id=message_id,
            netconf_namespace=netconf_namespace, xmlns=xmlns,
            calls=calls, templates=templates, kwargs=kwargs,
            strict_check=strict_check)
    finally:
        # goodbye
        ctx.logger.info("Connection close")
        message_id += 1
        goodbye_string = _generate_goodbye(
            xmlns, netconf_namespace, message_id
        )
        ctx.logger.debug("Sent: {message}"
                         .format(message=filters.shorted_text(goodbye_string)))

        response = netconf.close(goodbye_string)
        ctx.logger.debug("Recieved: {message} "
                         .format(message=filters.shorted_text(response)))
